source/extensions/filters/common/mutation_rules/mutation_rules.cc

```diff
@@ -0,0 +1,89 @@
+#include "source/extensions/filters/common/mutation_rules/mutation_rules.h"
+
+#include "envoy/http/header_map.h"
+
+#include "source/common/protobuf/utility.h"
+
+#include "absl/strings/match.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace MutationRules {
+
+Checker::Checker(const envoy::config::common::mutation_rules::v3::HeaderMutationRules& rules)
+    : rules_(rules) {
+  if (rules.has_allow_expression()) {
+    allow_expression_ = Regex::Utility::parseRegex(rules.allow_expression());
+  }
+  if (rules.has_disallow_expression()) {
+    disallow_expression_ = Regex::Utility::parseRegex(rules.disallow_expression());
+  }
+}
+
+// Pre-populate "routing headers" that we'll use to match whether a header
+// should be rejected according to the "allow_routing" rule.
+absl::flat_hash_set<Http::LowerCaseString> Checker::createRoutingHeaders() {
+  absl::flat_hash_set<Http::LowerCaseString> routing_headers;
+  const auto& headers = Http::Headers::get();
+  routing_headers.insert(headers.HostLegacy);
+  routing_headers.insert(headers.Host);
+  routing_headers.insert(headers.Method);
+  routing_headers.insert(headers.Scheme);
+  return routing_headers;
+}
+
+const absl::flat_hash_set<Http::LowerCaseString>& Checker::getRoutingHeaders() {
```

Comment:
This can be done with the `CONSTRUCT_ON_FIRST_USE` macro.

Reply:
Thanks! Used the macro and followed the pattern.